[ENH] Provide instructions for federation API (#133)
neurobagel/documentation: docs/federate.md

> There are two main reasons to deploy local query federation:

Author: Lines 16ff just below mention that. We also talk about ways to not federate on the infrastructure page. Does that address your point or do you think we need an explicit section here when you should **not** use federation?

> There are two main reasons to deploy local query federation:

Reviewer: Hm, that's true. I guess I was thinking from the perspective of someone who is only interested in federated queries and jumps straight to this page.

One option could be to move lines 16-19 to the top of this page, so that a user can see upfront that, without having to set up anything locally, they can already run federated queries over Neurobagel's public datasets. Wdyt?

> There are two main reasons to deploy local query federation:

Author: Yeah, I think I would do that as a "contraindications" section

> - **Case 1**: read-only federation. You have (at least) one [local neurobagel

Author: Agree on need to find better name. I want the name to very strongly indicate that under no circumstances can any of your data be seen from the outside / it's a one-way federation (maybe that's the name?). So semi-private/closed would not work here.

> - **Case 1**: read-only federation. You have (at least) one [local neurobagel

Reviewer: Oh, I see!

I think “one-way federation” describes the use case a lot more clearly - I vote for that! :)

node](infrastructure.md) and you want your users to be able to search
the data in the local node alongside all the publicly
visible data in the neurobagel network.
- **Case 2**: internal federation. You have two or more local neurobagel
nodes (e.g. for data from different groups in your institute)
and you want your local users to search across both of them.

![Local federation scenarios](imgs/local_federation_architecture.jpg)

Note that these cases are not mutually exclusive. 
Any local neurobagel nodes you deploy will only be visible to users
inside of your local network (internal federation).
Existing public nodes in the neurobagel network are accessible
to everyone (e.g. on [query.neurobagel.org](https://query.neurobagel.org/))
and you can choose to also add them to your local deployment
through read-only federation.

## Setting up for local federation

> Local federation in neurobagel is provided by the federation API (`f-API`) service.

Author: probably need to add the "local" part back in - but I'll apply your suggestion first

The neurobagel `f-API` takes a single user query and then sends it to every

> neurobagel node API (`n-API`) it is aware of, collects and combinesthe responses,

Author: disagree here. It's more accurate but also less readable. n-API subset selection is more of a concern for the query tool I'd say

> neurobagel node API (`n-API`) it is aware of, collects and combinesthe responses,

Author: alt: we could see if mkdocs materials has footnotes

> neurobagel node API (`n-API`) it is aware of, collects and combinesthe responses,

Reviewer: > It's more accurate but also less readable. n-API subset selection is more of a concern for the query tool I'd say

I think that’s fair! In that case, I’m okay to leave that detail out.

and sends them back to the user as a single answer.

!!! note

    Make sure you have at least one [local `n-API` configured and running](infrastructure.md)
    before you set up local federation. If you do not have any local
    `n-APIs` to federate over, you can just use the public [query.neurobagel.org](https://query.neurobagel.org/)
    query tool directly

In your command line, navigate to a new directory where you will keep the configuration
files for your new `f-API`. In this directory, create two files:

> **`fed.env` environment file**. 

Author: I think it does, and I like your alternative

Create a new textfile to hold environment variables needed for the `f-API` deployment. 

> Let's assume you are running two local nodes:

Author: we need an example with two n-APIs on different machines. Maybe 
```suggestion
Let's assume there are two local nodes already running on different servers of your institutional network, and you want to set up federation across both nodes:
```

- a node named `"node_archive"` running on your local computer on port `8000` and 

> - a node named `"node_recruitment"` running on a different computer with the local IP `192.168.0.1`, listening on the default http port `80`. 

Author: local IP here means that those are IPs in the local / private network of the institute. Is there something else that's confusing?

> - a node named `"node_recruitment"` running on a different computer with the local IP `192.168.0.1`, listening on the default http port `80`. 

Reviewer: Aha, I understand. I think I assume “local IP” means localhost/127.0.0.1, but I’m realizing that’s inaccurate to what the term is actually used to refer to.

Let’s stick to your original wording. :)

> In your `fed.env` file you would configure this as follows:

Author: I think that's an option, we could even highlight the line that needs adding. But maybe let's first have a quick chat about https://github.com/neurobagel/api/issues/162 and see if we will soon(ish) move towards service specific config files and how that might affect our defaults

``` {.bash .annotate}
# Configuration for f-API
# List of known local node APIs: (node_URL, node_NAME)
LOCAL_NB_NODES=(http://localhost:8000, node_archive) (http://192.168.0.1, node_recruitment)

> # Chose the port that the f-API will listen on (default 8000)
> NB_API_PORT=8080

Author: ```suggestion
# Define the port that the f-API will run on INSIDE the docker container (default 8000)
NB_API_PORT=8000
# Define the port that the f-API will be exposed on to the host computer (and likely the outside network)
NB_API_PORT_HOST=8080
```

# Chose the docker image tag of the f-API (default latest)
NB_API_TAG=latest

# Configuration for query tool
# Define the URL of the f-API as it will appear to a user

> API_QUERY_URL=http:localhost:8080 # (1)!

Author: You mean that it should correspond to an exposed port on the server rather to whatever port the API is running on inside the docker container on the server? 

I'm not sure how much sys-admin knowledge to put in here with the docs that are specific to our tools. Maybe a little info box? E.g. if we run NGINX (on 80/443) at that IP, we wouldn't specify a port either, and so on...

> # Chose the docker image tag of the query tool (default latest)
> NB_QUERY_TAG=latest

Author: ```suggestion
# Chose the docker image tag of the query tool (default latest)
NB_QUERY_TAG=latest
# Chose the port that the query tool will be exposed on the host and likely the network (default 3000)
NB_QUERY_PORT_HOST=3000
```

```

1.  When a user users the graphical query tool to query your
    f-API, these requests will be sent from the users machine,
    not from the machine hosting the query tool.

    Make sure you set the `API_QUERY_URL` in your `fed.env`
    as it will appear to a user on their own machine 
    - otherwise the request will fail..

> 

Author: Hhm, I think if we want to provide more detail here, then we should have a worked example, i.e. if API is on machine A and user's browser on machine B...

If you think that's important I can add that.

Each node is described in `LOCAL_NB_NODES` by a comma-delimited tuple of the form `(node_URL, node_NAME)`.

You can add one or more local nodes to the list of nodes known to your `f-API` in this way.
Just adjust the above snippets to your own deployment, and store it in a file called `fed.env`.

**`docker-compose.yml` docker config file**.

Create a second textfile called `docker-compose.yml`. 
This file describes the required services, ports and paths
to launch the `f-API` together with a connected query tool.

!!! danger "Make sure you have a recent version of docker compose installed"

    Some Linux distributions come with outdated versions of `docker` and 
    `docker compose` installed. Please make sure you install `docker` 
    as described in the [official documentation](https://docs.docker.com/engine/install/).

> Copy the following snippet into your `docker-compose.yml` file.

Author: I think that's something to discuss in https://github.com/neurobagel/api/issues/162. But yeah, I agree!

You should not have to change anything about this file.
All local configuration changes are done in the `fed.env` file.

``` {.yaml .annotate}
version: "3.8"

services:
  federation:
    image: "neurobagel/federation_api:${NB_API_TAG:-latest}"
    ports:

>       - "${NB_API_PORT_HOST:-8000}:${NB_API_PORT:-8000}"

Author: Aha aha, great catch, guess I confused myself here.

    env_file:
      - federate.env
    environment:
      - LOCAL_NB_NODES=${LOCAL_NB_NODES} # (1)!
      - NB_API_PORT=${NB_API_PORT:-8000}
  query:
    image: "neurobagel/query_tool:${NB_QUERY_TAG:-latest}"
    ports:
      - "${NB_QUERY_PORT_HOST:-3000}:3000"
    env_file:
      - federate.env
    environment:
      - API_QUERY_URL=${API_QUERY_URL:-http://localhost:8000/}
```

1.  We maintain a list of public neurobagel nodes 
    [here](https://github.com/neurobagel/menu/blob/main/node_directory/neurobagel_public_nodes.json).
    By default every new `f-API` will lookup this list
    on startup and include it in the list of nodes to
    federate over.
    This also means that you do not have to manually

>     configure public nodes.

Author: ```suggestion
    configure public nodes, i.e. you **do not have to explicitly add them** to the `LOCAL_NB_NODES` variable) in your `fed.env` file.
```

>     configure public nodes.

Reviewer: Looks good to me!

## When to use local query federation
There are two main reasons to deploy local query federation:

- **Case 1**: one-way federation. You have (at least) one [local neurobagel
node](infrastructure.md) and you want your users to be able to search
the data in the local node alongside all the publicly
visible data in the neurobagel network.
- **Case 2**: internal federation. You have two or more local neurobagel
nodes (e.g. for data from different groups in your institute)
and you want your local users to search across all of them.

![Local federation scenarios](imgs/local_federation_architecture.jpg)

Note that these cases are not mutually exclusive. 
Any local neurobagel nodes you deploy will only be visible to users
inside of your local network (internal federation).

## When not to use local query federation
Query federation is not necessary, if you:

- **only want to query public neurobagel nodes**:
  Existing public nodes in the neurobagel network are accessible

>   to everyone via our public query tool (e.g. on [query.neurobagel.org](https://query.neurobagel.org/))
>   and you can choose to also make them available to your local deployment
>   through one-way federation.

Reviewer: 🍒 :
```suggestion
  to everyone via our public query tool (e.g. on [query.neurobagel.org](https://query.neurobagel.org/)),
  meaning you can run federated queries over these graph databases without any additional local setup.
```
To avoid confusion, I wouldn't mention one-way federation here as it's not relevant to the case of "When **not** to use local query federation"
